core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
-      // We assume here that all clients use the same configuration.
-      ConfUtils.mergeHadoopConfiguration(conf);
-      Configuration.set(PropertyKey.MASTER_HOSTNAME, uri.getHost());
-      Configuration.set(PropertyKey.MASTER_RPC_PORT, Integer.toString(uri.getPort()));
-      Configuration.set(PropertyKey.ZOOKEEPER_ENABLED, Boolean.toString(isZookeeperMode()));
-
-      // These must be reset to pick up the change to the master address.
-      // TODO(andrew): We should reset key value system in this situation - see ALLUXIO-1706.
-      ClientContext.init();
-      sFileSystemContext.reset();
-      LineageContext.INSTANCE.reset();
-
-      // Try to connect to master, if it fails, the provided uri is invalid.
-      FileSystemMasterClient client = sFileSystemContext.acquireMasterClient();
-      try {
-        client.connect();
-        // Connected, initialize.
-        sFileSystem = FileSystem.Factory.get();
-        sInitialized = true;
-      } catch (ConnectionFailedException | IOException e) {
-        LOG.error("Failed to connect to the provided master address {}: {}.",
-            uri.toString(), e.toString());
-        throw new IOException(e);
-      } finally {
-        sFileSystemContext.releaseMasterClient(client);
-      }
+      initializeInternal(uri, conf);
+      sInitialized = true;
     }
+
+    setFileSystemAndContext();
   }
 
-  // Assures the mUri is the same as the master address in the current client context.
-  private void checkMasterAddress() throws IOException {
-    InetSocketAddress masterAddress = ClientContext.getMasterAddress();
+  /**
+   * Initializes the default contexts if the master address specified in the URI is different
+   * from the default one.
+   *
+   * @param uri the uri
+   * @param conf the hadoop conf
+   * @throws IOException if it fails to initialize
+   */
+  void initializeInternal(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
+    // Load Alluxio configuration if any and merge to the one in Alluxio file system. These
+    // modifications to ClientContext are global, affecting all Alluxio clients in this JVM.
+    // We assume here that all clients use the same configuration.
+    ConfUtils.mergeHadoopConfiguration(conf);
+    Configuration.set(PropertyKey.MASTER_HOSTNAME, uri.getHost());
+    Configuration.set(PropertyKey.MASTER_RPC_PORT, Integer.toString(uri.getPort()));
+    Configuration.set(PropertyKey.ZOOKEEPER_ENABLED, Boolean.toString(isZookeeperMode()));
+
+    // These must be reset to pick up the change to the master address.
+    // TODO(andrew): We should reset key value system in this situation - see ALLUXIO-1706.
+    LineageContext.INSTANCE.reset();
+    FileSystemContext.INSTANCE.reset();
+
+    // Try to connect to master, if it fails, the provided uri is invalid.
+    FileSystemMasterClient client = FileSystemContext.INSTANCE.acquireMasterClient();
```

Comment:
I think these two are equivalent. I prefer not to use "try-with-resource" because I need to mock the return value of "FileSystemContext.INSTANCE.acquireMasterClientResource()". 
check AbstractFileSystemTest#mockFileSystemContextAndMasterClient.